File: code/416/Java/dp.java
class Solution {
    public boolean canPartition(int[] nums) {
        // int total = IntStream.of(nums).sum();
        int total = 0;
        for (int n:nums) {
            total += n;
        }
        if ((total&1)==1) {
            return false;
        }
        int target = total>>1;
        boolean dp[] = new boolean[target+1];
        dp[0] = true;
        for (int n:nums) {
            for (int i=target; i>=n; i--) {
                dp[i] |= dp[i-n];
            }
        }
        return dp[target];
    }
}

File: code/416/Python/dp.py
class Solution:
    def canPartition(self, nums):
        total = sum(nums)
        if total&1: return False
        target = total>>1
        dp = [False]*(target+1)
        dp[0] = True
        for n in nums:
            for i in range(target, n-1, -1):
                dp[i] |= dp[i-n]
        return dp[target]